### PR TITLE
Support multiple ServicePorts per Service

### PR DESCRIPTION
**Pull request by falzm**
This change is basically a global rewrite of the LoadBalancer-related
section of the CCM in order to natively support the typical use case
of multiple k8s `ServicePorts` defined in a `Service`, e.g. for an
Ingress Controller that exposes both ports 80 and 443 for the same
service.

Also, this change replaces most of the shellscripting related to
setting up then tearing down the cluster infrastructure required for the
integration tests by a Terraform configuration, resulting in a much more
manageable and reliable setup.

Finally, the documentation has been updated according to the changes
described above.

### CODE REVIEW

File: integtest/manifests/hello.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    app: hello
+spec:
+  rules:
+  - http:
+      paths:
+      - path: /
+        pathType: Prefix
+        backend:
+          serviceName: hello
+          servicePort: 80
```
Comment (illi-j) [inline integtest/manifests/hello.yml:51]: newline ;)

Comment (falzm) [reply integtest/manifests/hello.yml]: Fixed :)

File: integtest/terraform.tfvars [hunk 2]
```diff
@@ -0,0 +1,2 @@
+zone = "de-fra-1"
+tmpdir = "/Users/marc/Documents/Exoscale/git/exoscale-cloud-controller-manager/integtest/tmp"
```
Comment (illi-j) [inline integtest/terraform.tfvars:2]: is this expected ?

File: integtest/test-helpers.bash [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  if [[ ! "$v" =~ $pattern ]]; then
+      echo "error: \"$v\" doesn't match /$pattern/"
+#      exit 1
+  fi
+}
+
+_until_success() {
+    declare command="$1"
+    timeout --preserve-status 600s bash -c "until $command > /dev/null 2>&1; do sleep 5; done"
+}
```
Comment (illi-j) [inline integtest/test-helpers.bash:26]: newline

File: integtest/test-labels.bash [hunk 4]
```diff
@@ -1,28 +1,32 @@
...
+  k=${l%=*} v=${l#*=}
+
+  # Ignore labels not managed by the Exoscale CCM
+  [[ -z "${EXPECTED[$k]}" ]] && continue
+
+  if [[ "$v" != "${EXPECTED[$k]}" ]]; then
+    echo "FAIL: Node label $k: expected \"${EXPECTED[$k]}\", got \"$v\""
+    exit 1
+  fi
+done < "${INTEGTEST_TMP_DIR}/nodepool_labels"
```
Comment (illi-j) [inline integtest/test-labels.bash:32]: same here ;)

File: integtest/terraform.tfvars (hunk 2 above)
Comment (falzm) [reply integtest/terraform.tfvars]: Well... no, good catch 😅

File: integtest/test-helpers.bash (hunk 3 above)
Comment (falzm) [reply integtest/test-helpers.bash]: Fixed 👍

File: integtest/test-labels.bash (hunk 4 above)
Comment (falzm) [reply integtest/test-labels.bash]: Fixed 👍

**Review by illi-j — APPROVED** [review]
LGTM! 👍  thx @falzm

File: exoscale/loadbalancer.go [hunk 5]
```diff
@@ -109,154 +97,193 @@ func (l *loadBalancer) EnsureLoadBalancer(ctx context.Context, _ string, service
...
+func (l *loadBalancer) EnsureLoadBalancerDeleted(ctx context.Context, _ string, service *v1.Service) error {
 	zone, err := l.getLoadBalancerZone(service)
 	if err != nil {
 		return err
 	}
 
-	lbDef, err := buildLoadBalancerFromAnnotations(service)
+	nlb, err := l.fetchLoadBalancer(ctx, service, zone)
 	if err != nil {
+		if err == errLoadBalancerNotFound {
```
Comment (pierre-emmanuelJ) [inline exoscale/loadbalancer.go:171]: If you want to stay consistent with your previous code, you can use
`errors.Is()` function.

File: exoscale/loadbalancer.go [hunk 6]
```diff
@@ -109,154 +97,193 @@ func (l *loadBalancer) EnsureLoadBalancer(ctx context.Context, _ string, service
...
+		for _, nlbServiceUpdate := range nlbUpdate.Services {
+			// If a service exposing the same port already exists,
+			// flag it for update and save its ID for later reference.
+			if nlbServiceUpdate.Port == nlbServiceCurrent.Port {
+				debugf("service port %d already in use by NLB service %q, marking for update",
+					nlbServiceCurrent.Port,
+					nlbServiceCurrent.Name)
+				nlbServices[nlbServiceCurrent.Port] = nlbServiceCurrent
+				continue next
+			}
```
Comment (pierre-emmanuelJ) [inline exoscale/loadbalancer.go:242]: This code could be spot into a function, instead of continuing on the  `next` keyword.

File: go.sum [hunk 7]
```diff
@@ -561,6 +561,8 @@ github.com/stretchr/testify v1.4.0 h1:2E4SXV/wtOkTonXsotYi4li6zVWxYlZuYNCXe9XRJy
 github.com/stretchr/testify v1.4.0/go.mod h1:j7eGeouHqKxXV5pUuKE4zz7dFj8WfuZ+81PSLYec5m4=
 github.com/stretchr/testify v1.5.1 h1:nOGnQDM7FYENwehXlg/kFVnos3rEvtKTjRvOWSzb6H4=
 github.com/stretchr/testify v1.5.1/go.mod h1:5W2xD1RspED5o8YsWQXVCued0rvSQ+mT+I5cxcmMvtA=
+github.com/stretchr/testify v1.6.1 h1:hDPOHmpOpP40lSULcqw7IrRb/u7w6RpDC9399XyoNd0=
```
Comment (pierre-emmanuelJ) [inline go.sum:564]: Could you run a `go mod tidy`, thanks.

Comment (falzm) [reply go.sum:564]: I did – different versions are referenced in the vendored modules.

File: exoscale/loadbalancer.go (hunk 5 above)
Comment (falzm) [reply exoscale/loadbalancer.go]: Good catch, fixed 👍

File: exoscale/loadbalancer.go (hunk 6 above)
Comment (falzm) [reply exoscale/loadbalancer.go:242]: @pierre-emmanuelJ can you show me how you'd do it with a function?

**Review by pierre-emmanuelJ — APPROVED** [review]

**Pull request merged**